seal5/logging.py

```diff
@@ -21,76 +21,176 @@
+    
+    server_reachable = False
+    try:
+        with socket.create_connection((HOSTNAME, SEAL5_LOGGING_PORT), timeout=0.2):
+            server_reachable = True
+    except OSError:
+        server_reachable = False
+
+    # --- Case 1: Logging server not reachable → local fallback logger fallback ---
+    if not server_reachable:
+        fallback_logger = logging.getLogger('fallback')  # fallback logger
+        fallback_logger.setLevel(logging.DEBUG)
+
+        # Ensure a StreamHandler exists only once
+        if not any(isinstance(h, logging.StreamHandler) for h in fallback_logger.handlers):
+            stream_handler = logging.StreamHandler(sys.stdout)
+            stream_handler.setFormatter(get_formatter(True))
+            stream_handler.setLevel(logging.DEBUG)
+            fallback_logger.addHandler(stream_handler)
+
+        return fallback_logger
+
+    logger = logging.getLogger(f"{PROJECT_NAME}.{loggername if loggername is not None else 'unknown'}")
+    logger.handlers = []
+    logger.setLevel(level=level)
+    socket_handler = logging.handlers.SocketHandler(HOSTNAME, SEAL5_LOGGING_PORT)
+    logger.addHandler(socket_handler)
     return logger
 
 
-def set_log_level(console_level=None, file_level=None):
-    """Set command line log level at runtime."""
-    # print("set_log_level", console_level, file_level)
-    logger = logging.getLogger("seal5")
-    for handler in logger.handlers[:]:
-        # print("handler", handler, type(handler))
-        if (
-            isinstance(handler, (logging.FileHandler, logging.handlers.TimedRotatingFileHandler))
-            and file_level is not None
-        ):
-            handler.setLevel(file_level)
-            # print("NEWIF")
-        elif isinstance(handler, logging.StreamHandler) and console_level is not None:
-            handler.setLevel(console_level)
-            # print("NEWELIF")
+def initialize_logging_server(
+    logfiles: None | List[Tuple[Path | str, int]] = [
+        ("log_debug.log", logging.DEBUG),
+        ("log_info.log", logging.INFO),
+    ], stream_log_level: int | str = logging.INFO,
+):
+    global _log_server
+    logger = logging.getLogger(PROJECT_NAME)
+    # This should be the lowest value and not changeable since logger is the first filter
     logger.setLevel(logging.DEBUG)
 
+    stream_handler = logging.StreamHandler(sys.stdout)
+    stream_handler.setFormatter(get_formatter(True))
+    stream_handler.setLevel(resolve_log_level(stream_log_level))
+    logger.addHandler(stream_handler)
+
+    if logfiles is not None:
+        for log_file, log_level in logfiles:
+            file_handler = logging.FileHandler(log_file, "w")
+            file_handler.setFormatter(get_formatter(False))
+            file_handler.setLevel(resolve_log_level(log_level))
+            logger.addHandler(file_handler)
+
+    try:
+        _log_server = LogRecordSocketReceiver()
+    except OSError as e:
+        if e.errno == 98:  # Address already in use
```

Comment:
I often have multiple seal5 session running in parallel. We should probably look into a way to automatically pick a free port...